src/uffs/uffs_fs.c

```diff
@@ -557,6 +557,71 @@ URET uffs_OpenObjectEx(uffs_Object *obj, uffs_Device *dev,
 	return (obj->err == UENOERR ? U_SUCC : U_FAIL);
 }
 
+URET uffs_OpenObjectBySerial(uffs_Object *obj, uffs_Device *dev, int serial, int oflag)
+{
+		obj->err = UENOERR;
+	obj->open_succ = U_FALSE;
+
+	if (dev == NULL) {
+		obj->err = UEINVAL;
+		goto ext;
+	}
+
+	if ((oflag & (UO_WRONLY | UO_RDWR)) == (UO_WRONLY | UO_RDWR)) {
+		/* UO_WRONLY and UO_RDWR can't appear together */
+		uffs_Perror(UFFS_MSG_NOISY,
+					"UO_WRONLY and UO_RDWR can't appear together");
+		obj->err = UEINVAL;
+		goto ext;
+	}
+
+	obj->dev = dev;
+	obj->type = (oflag & UO_DIR ? UFFS_TYPE_DIR : UFFS_TYPE_FILE);
+
+	uffs_ObjectDevLock(obj);
+
+	if (obj->type == UFFS_TYPE_DIR) {
+		obj->node = uffs_TreeFindDirNode(obj->dev, serial);
+	}
+	else {
+		obj->node = uffs_TreeFindFileNode(obj->dev, serial);
+	}
+
+	if (obj->node == NULL) {			// dir or file not exist
+		obj->err = UENOENT;
+		goto ext_1;
+	}
+
+	obj->oflag = oflag;
+	obj->parent = GET_OBJ_NODE_FATHER(obj);
+	obj->pos = 0;
+	obj->name = 0;
+	obj->name_len = 0;
+
+	obj->sum = 0;
+	obj->head_pages = obj->dev->attr->pages_per_block - 1;
+
+	if ((obj->oflag & (UO_CREATE | UO_EXCL)) == (UO_CREATE | UO_EXCL)){
+		obj->err = UEEXIST;
+		goto ext_1;
+	}
+
+	obj->serial = GET_OBJ_NODE_SERIAL(obj);
+	obj->open_succ = U_TRUE;
+
+	if (obj->oflag & UO_TRUNC)
```

Comment:
I know that original code didn't have brackets for this `if` but maybe lets fix that in here.